File: src/model/Time.java
package model;

import java.util.Random;

public class Time {

	int time, day, year;
	Month month;
	
	private enum TimeOfDay {
		MORNING,
		AFTERNOON,
		EVENING,
		NIGHT;
	}
	
	private enum Month {
		JANUARY(31, "Jan."),
		FEBRUARY(28, "Feb."),
		MARCH(31, "March"),
		APRIL(30, "April"),
		MAY(31, "May"),
		JUNE(30, "June"),
		JULY(30, "July"),
		AUGUST(31, "Aug."),
		SEPTEMBER(30, "Sept."),
		OCTOBER(31, "Oct."),
		NOVEMBER(30, "Nov."),
		DECEMBER(31, "Dec.");
		
		private int numberOfDays;
		private String name;
		
		private Month(int numberOfDays, String name) {
			this.numberOfDays = numberOfDays;
			this.name = name;
		}
		
		public int getNumberOfDays() {
			return numberOfDays;
		}
		
		public String getName() {
			return name;
		}
		public Month getNextMonth() {
			return values()[(ordinal() + 1) % (values().length)];
		}
	}
	
	
	public Time() {
		Random random = new Random();
		this.time = random.nextInt(24);
		this.month = Month.values()[random.nextInt(Month.values().length)];
		this.day = random.nextInt(month.getNumberOfDays() - 1) + 1;
		this.year = random.nextInt(10) + 1860;
	}
	public Time(int time, int day, int month, int year) {
		this.time = time;
		this.day = day;
		this.month = Month.values()[month - 1];
		this.year = year;
	}
	
	public TimeOfDay getTimeOfDay() {
		if( 7 <= time && time <= 12) {
			return TimeOfDay.MORNING;
		} else if ( 13 <= time && time <= 18) {
			return TimeOfDay.AFTERNOON;
		} else if (19 <= time || time <= 0) {
			return TimeOfDay.EVENING;
		} else {
			return TimeOfDay.NIGHT;
		}
	}
	
	public void advanceTime() {
		time = (time + 1);
		if ( time == 23) {
			day += 1;
			if (day > month.getNumberOfDays() + (month.equals(Month.FEBRUARY) && 
				((year%4 == 0 && year%100 != 0) || (year%400) == 0) ? 1 : 0)) {
				day = 1;
				month = month.getNextMonth();
				if (month == Month.JANUARY) {
					year += 1;
				}
			}
		} else if (time == 24) {
			time = 0;
		}
	}

	public int getTime() {
		return time;
	}
	
	public String get24HourTime() {
		return time + ":00";
	}
	
	public String get12HourTime() {
		return ((time)%12 + 1) + ":00" + ((time < 11 || time == 23)? "am" : "pm");
	}
	
	public String getDayMonthYear() {
		return month.getName() + " " + day + ", " + year;
	}
}

File: web/src/model/Time.ts
enum Month {
  JANUARY = 0,
  FEBRUARY = 1,
  MARCH = 2,
  APRIL = 3,
  MAY = 4,
  JUNE = 5,
  JULY = 6,
  AUGUST = 7,
  SEPTEMBER = 8,
  OCTOBER = 9,
  NOVEMBER = 10,
  DECEMBER = 11,
}

const MONTH_DATA: Record<Month, { numberOfDays: number; name: string }> = {
  [Month.JANUARY]: { numberOfDays: 31, name: "Jan." },
  [Month.FEBRUARY]: { numberOfDays: 28, name: "Feb." },
  [Month.MARCH]: { numberOfDays: 31, name: "March" },
  [Month.APRIL]: { numberOfDays: 30, name: "April" },
  [Month.MAY]: { numberOfDays: 31, name: "May" },
  [Month.JUNE]: { numberOfDays: 30, name: "June" },
  [Month.JULY]: { numberOfDays: 30, name: "July" },
  [Month.AUGUST]: { numberOfDays: 31, name: "Aug." },
  [Month.SEPTEMBER]: { numberOfDays: 30, name: "Sept." },
  [Month.OCTOBER]: { numberOfDays: 31, name: "Oct." },
  [Month.NOVEMBER]: { numberOfDays: 30, name: "Nov." },
  [Month.DECEMBER]: { numberOfDays: 31, name: "Dec." },
};

export enum TimeOfDay {
  MORNING = "MORNING",
  AFTERNOON = "AFTERNOON",
  EVENING = "EVENING",
  NIGHT = "NIGHT",
}

export class Time {
  private time: number;
  private day: number;
  private month: Month;
  private year: number;

  constructor(time?: number, day?: number, month?: number, year?: number) {
    if (time !== undefined && day !== undefined && month !== undefined && year !== undefined) {
      this.time = time;
      this.day = day;
      this.month = (month - 1) as Month;
      this.year = year;
    } else {
      this.time = Math.floor(Math.random() * 24);
      this.month = Math.floor(Math.random() * 12) as Month;
      const daysInMonth = MONTH_DATA[this.month].numberOfDays;
      this.day = Math.floor(Math.random() * (daysInMonth - 1)) + 1;
      this.year = Math.floor(Math.random() * 10) + 1860;
    }
  }

  getTimeOfDay(): TimeOfDay {
    if (this.time >= 7 && this.time <= 12) {
      return TimeOfDay.MORNING;
    } else if (this.time >= 13 && this.time <= 18) {
      return TimeOfDay.AFTERNOON;
    } else if (this.time >= 19 || this.time <= 0) {
      return TimeOfDay.EVENING;
    } else {
      return TimeOfDay.NIGHT;
    }
  }

  advanceTime(): void {
    this.time = this.time + 1;

    if (this.time === 23) {
      this.day += 1;
      const isLeapYear = (this.year % 4 === 0 && this.year % 100 !== 0) || this.year % 400 === 0;
      const daysInMonth = MONTH_DATA[this.month].numberOfDays + (this.month === Month.FEBRUARY && isLeapYear ? 1 : 0);

      if (this.day > daysInMonth) {
        this.day = 1;
        this.month = ((this.month + 1) % 12) as Month;
        if (this.month === Month.JANUARY) {
          this.year += 1;
        }
      }
    } else if (this.time === 24) {
      this.time = 0;
    }
  }

  getTime(): number {
    return this.time;
  }

  get24HourTime(): string {
    return `${this.time}:00`;
  }

  get12HourTime(): string {
    const hour = (this.time % 12) + 1;
    const period = this.time < 11 || this.time === 23 ? "am" : "pm";
    return `${hour}:00${period}`;
  }

  getDayMonthYear(): string {
    return `${MONTH_DATA[this.month].name} ${this.day}, ${this.year}`;
  }
}
